handlers/aws/utils.py

```diff
@@ -125,13 +125,93 @@ def wrapper(lambda_event: dict[str, Any], lambda_context: context_.Context) -> s
             if apm_client:
                 apm_client.capture_exception()
 
-            shared_logger.exception("exception raised", exc_info=e)
+            shared_logger.exception(
+                "exception raised",
+                exc_info=e,
+                extra={
+                    "summary": summarize_lambda_event(lambda_event, max_records=20),
+                },
+            )
 
             return f"exception raised: {e.__repr__()}"
 
     return wrapper
 
 
+def summarize_lambda_event(event: dict[str, Any], max_records: int = 10) -> dict[str, Any]:
+    """
+    Summarize the lambda event to include only the most relevant information.
+    """
+    summary: dict[str, Any] = {}
+
+    try:
+        first_records_key = f"first_{max_records}_records"
+        records = event.get("Records", [])
+
+        for record in records:
+            event_source = record.get("eventSource", "unknown")
+
+            if event_source == "aws:sqs":
+                aws_sqs_summary = summary.get(
+                    "aws:sqs",
+                    # if `aws:sqs` key does not exist yet,
+                    # we initialize the summary.
+                    {
+                        "total_records": 0,
+                        first_records_key: [],
+                    },
+                )
+
+                # We keep track of the total number of notifications in the
+                # lambda event, so users know if the summary is incomplete.
+                notifications = json_parser(record["body"])
+
+                # So users know if we included only a
+                # subset of the records.
+                aws_sqs_summary["total_records"] += len(notifications["Records"])
+
+                for r in notifications["Records"]:
+                    # we only include the s3 object key in the summary.
+                    #
+                    # Here is an example of a record:
+                    #
+                    # {
+                    #   "Records": [
+                    #     {
+                    #       "awsRegion": "eu-west-1",
+                    #       "eventName": "ObjectCreated:Put",
+                    #       "eventSource": "aws:s3",
+                    #       "eventVersion": "2.1",
+                    #       "s3": {
+                    #         "bucket": {
+                    #           "arn": "arn:aws:s3:::mbranca-esf-data",
+                    #           "name": "mbranca-esf-data"
+                    #         },
+                    #         "object": {
+                    #           "key": "AWSLogs/1234567890/CloudTrail-Digest/"
+                    #         }
+                    #       }
+                    #     }
+                    #   ]
+                    # }
+
+                    # We stop adding records to the summary once we reach
+                    # the `max_records` limit.
+                    if len(aws_sqs_summary[first_records_key]) < max_records:
```

Comment:
Should you also break the loop when reaching that? If so you can reverse the check and add a break to avoid looping for nothing.

Reply:
I was continuing the loop to get the total number of records.

Here's the lambda event structure (with minor changes for readability):

```json
{
    "Records": [
        {
            "messageId": "059f36b4-87a3-44ab-83d2-661975830a7d",
            "receiptHandle": "AQEBwJnKyrHigUMZj6rYigCgxlaS3SLy0a...",
            "body": "{\"Records\":[{\"awsRegion\":\"eu-west-1\",\"eventTime\":\"2024-01-31T19:06:30.680Z\",\"eventName\":\"ObjectCreated:Put\",\"eventSource\":\"aws:s3\",\"eventVersion\":\"2.1\",\"s3\":{\"bucket\":{\"arn\":\"arn:aws:s3:::mbranca-esf-data\",\"name\":\"mbranca-esf-data\"},\"object\":{\"key\":\"AWSLogs/123456/I-do-not-exist\"}}}]}",
            "attributes": {
                "ApproximateReceiveCount": "1",
                "SentTimestamp": "1545082649183",
                "SenderId": "AIDAIENQZJOLO23YVJ4VO",
                "ApproximateFirstReceiveTimestamp": "1545082649185"
            },
            "messageAttributes": {
                "myAttribute": {
                    "stringValue": "myValue", 
                    "stringListValues": [], 
                    "binaryListValues": []
                }
            },
            "md5OfBody": "e4e68fb7bd0e697a0ae8f1bb342846b3",
            "eventSource": "aws:sqs",
            "eventSourceARN": "arn:aws:sqs:us-east-2:123456789012:my-queue",
            "awsRegion": "us-east-2"
        }
    ]
}
```

So my goal is to provide the total number of the following elements (hoppa `jq` style):

```text
.Records[].body.Records
```

Reply:
oh, I see you point now, we can skip the inner loop 🤦